[PSQ] Allow using Classic McEliece for PQ KEM

jschneider-bensch opened the pull request:
This PR does four things:
1. Clean up, some renaming and more documentation for  more of the new PSQ protocol. In particular, the doc test at the crate level showcases how a responder can be queried for its ciphersuite and how to use the response to perform a registration with the responder-supported ciphersuite. 
2. Move the old PSQ code to a feature-guarded `v1` module, with the exception of some of the Classic McEliece code.
3. Introduce an abstraction for PQ KEM keys, which allows to use either ML-KEM 768 or Classic McEliece 460896f as the PQ KEM.
4. Introduce names for all supported ciphersuites, and those that will be supported in the future (AES-GCM ones)

In terms of possible "configurations" for 3. I went the simplest option, i.e. an initiator can have either an ML-KEM or CMC longterm public key of the responder, and the responder can have either an ML-KEM or CMC key pair. If there is a mismatch, i.e. the initiator sends a registration message for which the responder does not have a matching decapsulation key, then the responder will ignore that message and return to its initial state.

**After this PR public-facing APIs should be stable**, I hope, with possible exception of changes from #1065 that I can't yet anticipate.

---
**TODO**
- [x] Ciphersuite builder
- [x] Test configuration mismatch

franziskuskiefer reviewed (commented):
Just some early thoughts

franziskuskiefer commented on libcrux-psq/benches/psq_v2.rs:
```diff
@@ -27,7 +30,7 @@ fn query<const PQ: bool>(c: &mut Criterion) {
     // External setup
     let responder_ecdh_keys = DHKeyPair::new(&mut rng);
-    let responder_pq_keys = PQKeyPair::new(&mut rng);
+    let responder_pq_keys = libcrux_ml_kem::mlkem768::rand::generate_key_pair(&mut rng);
```
Should the `into` once happen up here already?

franziskuskiefer commented on libcrux-psq/src/handshake/builder.rs:
```diff
@@ -0,0 +1,151 @@
+    }
+
+    /// Set the long-term PQ key pair.
+    pub fn longterm_pq_keys(mut self, longterm_pq_keys: impl Into<PQKeyPair<'a>>) -> Self {
```
We should make the API consistent. The DH above gets just a reference. We should pick a style and stick to it for both types of keys.

franziskuskiefer commented on libcrux-psq/tests/query.rs:
```diff
@@ -14,18 +17,18 @@ fn query() {
     // External setup
     let responder_ecdh_keys = DHKeyPair::new(&mut rng);
 
-    let responder_pq_keys = PQKeyPair::new(&mut rng);
+    let responder_pq_keys = libcrux_ml_kem::mlkem768::rand::generate_key_pair(&mut rng);
```
Same comment as for the builder below. We should pick a style for both. Here in the call site we just see the difference as well.
I think either could be fine. Either a generic `key_gen` with an algorithm identifier, or explicitly generating the keys. But I think I like one function for all a little better.

franziskuskiefer commented on libcrux-psq/src/handshake/builder.rs:
```diff
@@ -0,0 +1,151 @@
+        self
+    }
+
+    /// Set the inner AAD.
```
Somewhat unrelated to this PR but I think the aad setters need more documentation to tell the user what the outer and inner aads are.

franziskuskiefer commented on libcrux-psq/src/handshake/builder.rs:
```diff
@@ -0,0 +1,151 @@
+    /// is set.
+    /// It also uses the `context`, `inner_aad`, `outer_aad`, and
+    /// `peer_longterm_pq_pk`.
+    pub fn build_registration_initiator(self) -> Result<RegistrationInitiator<'a, Rng>, Error> {
```
I think there's no way to ensure that we set a pq key if we want a pq ciphersuite, right?
If the caller forgets to set the pq key, it'll just do dh only. That's not great.
That also goes back to the question about how to generate keys and ciphersuites.

I suggest that we add a ciphersuite in the `new` of the builder and that way enforce that the correct set of keys are available.
This allows then generating keys based on it. That's not so important though. It's long term keys and it's not clear whether that would actually be used.

jschneider-bensch replied:
I agree, it should be an explicit choice to use DH only!

franziskuskiefer reviewed (commented):
Some thoughts

jschneider-bensch commented on libcrux-psq/Cargo.toml:
```diff
@@ -46,17 +46,22 @@ log = "0.4.26"
 [features]
-classic-mceliece = ["dep:classic-mceliece-rust", "rand_old"]
+legacy = []
+# classic-mceliece = ["dep:classic-mceliece-rust", "rand_old"]
```
Yes, I temporarily disabled that too work on the new version as if CMC was unconditionally included, but in any case the old protocol needs the feature.

jschneider-bensch commented:
I've changed the proposal here to use a trait mechanism for implementing ciphersuites. This makes it easier to pull out Classic McEliece into a non-default feature.

The downside is that in the way I've implemented it here, we need to pre-serialize the PQ encapsulation, to make the `HandshakeMessage` types oblivious to the used ciphersuite and we need to do the same with the PQ encapsulation key in another place, since we don't yet have a blanket impl of `Serialize` for `Option<&'a T>` from `T: Serialize`. I tried getting around this by introducing a separate associated type for the reference to the encapsulation key, that would also require a `Serialize` bound, but could not get it to work. At some point I needed to go from `Option<C::EncapsulationKey>` to `Option<C::EncapsulationKeyRef>` and I couldn't figure out which (if any) bounds would allow me to do that.

One more thing that is missing here is a builder for the ciphersuites, so now they are statically constructed in the tests.

franziskuskiefer commented on libcrux-psq/src/handshake/ciphersuite.rs:
```diff
@@ -0,0 +1,747 @@
+#[repr(u8)]
+pub enum DynamicSharedSecret<'a> {
+    None,
+    MlKem(MlKemSharedSecret),
```
There are a couple clippy warnings, including the boxing for the enums.

franziskuskiefer commented on libcrux-psq/src/handshake/ciphersuite.rs:
```diff
@@ -0,0 +1,747 @@
+}
+
+impl<'a> CiphersuiteBuilder<'a> {
+    pub fn new() -> Self {
```
For builders you usually add a `builder()` function to the resulting object as well to make it more seamless. Maybe that's difficult here because of the different results. Also not too important. Just a note.

franziskuskiefer commented on libcrux-psq/src/handshake/ciphersuite.rs:
```diff
@@ -0,0 +1,747 @@
+
+    pub fn finish_responder(
+        self,
+        name: CiphersuiteName,
```
I'd expect the name in `new`. This is fine too. I was just wondering where the ciphersuite is.

franziskuskiefer commented on libcrux-psq/src/handshake/ciphersuite.rs:
```diff
@@ -0,0 +1,747 @@
+pub enum ResponderCiphersuite<'a> {
+    X25519_ChaChaPoly_HkdfSha256(ResponderX25519ChaChaPolyHkdfSha256<'a>),
+    X25519_MlKem768_ChaChaPoly_HkdfSha256(ResponderX25519MlKem768ChaChaPolyHkdfSha256<'a>),
+    #[cfg(feature = "classic-mceliece")]
```
Not the nicest, but I'm not sure what kind of trouble this may still be able to do. So I think this is fine. When everything is boxed it may also be equivalent in memory.

franziskuskiefer commented on libcrux-psq/src/handshake/ciphersuite.rs:
```diff
@@ -0,0 +1,747 @@
+            )),
+            #[cfg(not(feature = "classic-mceliece"))]
+            ResponderCiphersuite::X25519_ClassicMcEliece_ChaChaPoly_HkdfSha256(_) => {
+                panic!("unsupported ciphersuite")
```
Maybe let's return `None` in these cases and print an error + `debug_assert`. Just to avoid panics at runtime in deployments.
